tests/failpoint/db_failpoint_test.go

```diff
@@ -267,6 +270,97 @@ func TestIssue72(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestTx_Rollback_Freelist(t *testing.T) {
+	db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: 4096})
+
+	bucketName := []byte("data")
+
+	t.Log("Populate some data to have at least 5 leaf pages.")
+	var keys []string
+	err := db.Update(func(tx *bolt.Tx) error {
+		b, terr := tx.CreateBucket(bucketName)
+		if terr != nil {
+			return terr
+		}
+		for i := 0; i <= 10; i++ {
+			k := fmt.Sprintf("t1_k%02d", i)
+			keys = append(keys, k)
+
+			v := make([]byte, 1500)
+			if _, terr := crand.Read(v); terr != nil {
+				return terr
+			}
+
+			if terr := b.Put([]byte(k), v); terr != nil {
+				return terr
+			}
+		}
+		return nil
+	})
+	require.NoError(t, err)
+
+	t.Log("Remove some keys to have at least 3 more free pages.")
+	err = db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bucketName)
+		for i := 0; i < 6; i++ {
+			if terr := b.Delete([]byte(keys[i])); terr != nil {
+				return terr
+			}
+		}
+		return nil
+	})
+	require.NoError(t, err)
+
+	t.Log("Close and then reopen the db to release all pending free pages.")
+	db.MustClose()
+	db.MustReopen()
+
+	t.Log("Enable the `beforeWriteMetaError` failpoint.")
+	require.NoError(t, gofail.Enable("beforeWriteMetaError", `return("writeMeta somehow failed")`))
+
+	beforeFreelistPgids, err := readFreelistPageIds(db.Path())
+	require.NoError(t, err)
+
+	t.Log("Simulate TXN rollback")
+	err = db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bucketName)
+		for i := 6; i < len(keys); i++ {
+			v := make([]byte, 1500)
+			if _, terr := crand.Read(v); terr != nil {
+				return terr
+			}
+			// update the keys
+			if terr := b.Put([]byte(keys[i]), v); terr != nil {
+				return terr
+			}
+		}
+		return nil
+	})
+	require.Error(t, err)
+
+	afterFreelistPgids, err := readFreelistPageIds(db.Path())
+	require.NoError(t, err)
+
+	require.Equal(t, beforeFreelistPgids, afterFreelistPgids)
+
+	t.Log("Disable the `beforeWriteMetaError` failpoint.")
+	require.NoError(t, gofail.Disable("beforeWriteMetaError"))
```

Comment:
that's probably a better candidate for a `defer`

Reply:
If we use defer, then we have to remove the `require.NoError`, something like below,

```
defer gofail.Disable("beforeWriteMetaError")
```

Note each time a "defer" statement executes, the function value and parameters to the call are [evaluated as usual](https://go.dev/ref/spec#Calls) and saved anew but the actual function is not invoked.

Reply:
Updated again,

```
	defer func() {
		t.Log("Disable the `beforeWriteMetaError` failpoint.")
		require.NoError(t, gofail.Disable("beforeWriteMetaError"))
	}()

```